pkg/gconfig/gconfig.go

```diff
@@ -0,0 +1,232 @@
+	// This happens when Field.Dump(Dumper) is called with a secret dumper
+	secretUpdated bool
+	// secretPathPrefix defines the path that this secret should be written to.
+	// For example, in aws ssm, this is the secret path
+	secretPathPrefix string
+	// When a secret is read from file with the aws ssm loader, the path will be set here.
+	// If this is a newly created secret, when it is put in ssm, the path is saved here.
+	// this value is typically derived from the secretPathPrefix a suffix and a version number
+	secretPath string
+}
+
+func (s Field) HasChanged() bool {
+	return s.hasChanged
+}
+
+// Path returns the secret path
+// secrets loaded from config with the SSM Loader will have an secret path relevant to the loader type
+// secrets loaded from a test loader like JSONLoader or MapLoader will not have a path and this method will return an empty string
+func (s Field) SecretPath() string {
+	return s.secretPath
+}
+
+// IsSecret returns true if this Field is a secret
+func (s Field) IsSecret() bool {
+	return s.secret
+}
+
+// IsOptional returns true if this Field is optional
+func (s Field) IsOptional() bool {
+	return s.optional
+}
+
+// Key returns the key for this field
+func (s Field) Key() string {
+	return s.key
+}
+
+// Usage returns the usage string for this field
+func (s Field) Usage() string {
+	return s.usage
+}
+
+// Set the value of this string
+func (s *Field) Set(v string) error {
+	if s.value == nil {
+		return errors.New("cannot call Set on nil Valuer")
+	}
+	s.hasChanged = true
+	s.value.Set(v)
+	return nil
+}
+
+// Get returns the value if it is set, or an empty string if it is not set
+func (s *Field) Get() string {
+	if s.value == nil {
+		return ""
+	}
+	return s.value.Get()
+}
+
+// String calls the Valuer.String() method for this fields value.
+// If this field is a secret, then the response will be a redacted string.
+// Use Field.Get() to retrieve the raw value for the field
+func (s *Field) String() string {
+	if s.value == nil {
+		return ""
+	}
+	return s.value.String()
+}
+
+// SecretConfigValue value implements the Valuer interface, it should be used for secrets in configuration structs.
+//
+// It is configured to automatically redact the secret for common logging usecases like Zap, fmt.Println and json.Marshal
+type SecretConfigValue string
+
+// Get the raw value of the secret
+func (s *SecretConfigValue) Get() string {
+	if s == nil {
+		return ""
+	}
+	return string(*s)
+}
+
+// Set the value of the secret
+func (s *SecretConfigValue) Set(value string) {
+	*s = SecretConfigValue(value)
+}
+
+// String returns a redacted value for this secret
+func (s SecretConfigValue) String() string {
+	return "*****"
+}
+
+// MarshalJSON returns a redacted value bytes for this secret
+func (s SecretConfigValue) MarshalJSON() ([]byte, error) {
+	return json.Marshal(s.String())
+}
+
+// ConfigValue value implements the Valuer interface
+type ConfigValue string
```

Comment:
`ConfigValue` obscures the fact that it's a string under the hood. What about `StringValue`?

```suggestion
type StringValue string
```

Reply:
My original reasoning for this was a need to avoid a naming collision, but since then I renamed the underlying struct from Value to Field so I will update the Creator functions from String() and SecretString() to be Field() and SecretField()
Though I think I could event simplify it further by just having Field(SecretStringValue || StringValue) and infer whether its a secret from the value type

Reply:
Though that actually reduces readability so I will stick to SecretField() and Field()